sources/src/TorrentPlugin.cpp

```diff
@@ -164,67 +169,26 @@ void TorrentPlugin::on_piece_pass(int index) {
 
     // Make sure we are in correct mode, as mode changed may have occured
     if(_session.mode() == protocol_session::SessionMode::buying) {
-
-        auto it = _outstandingFullPieceArrivedCalls.find(index);
-
-        // If this validation is not due to us
-        if(it == _outstandingFullPieceArrivedCalls.cend()) {
-
-            // then just tell session about it
-            _session.pieceDownloaded(index);
-
-        } else {
-            auto peerId = it->second;
-            auto peerPlugin = peer(peerId);
-            auto endPoint = peerPlugin->endPoint();
-
-            _alertManager->emplace_alert<alert::ValidPieceArrived>(_torrent, endPoint, peerId, index);
-
-            // if its due to us, then tell session about endpoint and piece
-            _session.validPieceReceivedOnConnection(peerId, index);
-
-            // and remove call
-            _outstandingFullPieceArrivedCalls.erase(it);
-        }
+        _session.pieceDownloaded(index);
     }
 }
 
 void TorrentPlugin::on_piece_failed(int index) {
 
-    // Make sure we are in correct mode, as mode changed may have occured
-    if(_session.mode() == protocol_session::SessionMode::buying) {
-
-        auto it = _outstandingFullPieceArrivedCalls.find(index);
-
-        // If this validation is not due to us
-        if(it == _outstandingFullPieceArrivedCalls.cend()) {
-
-            // then there is nothing to do
-
-        } else {
-
-            // if its due to us, then
-
-            auto peerId = it->second;
-            auto peerPlugin = peer(peerId);
-            auto endPoint = peerPlugin->endPoint();
-
-            _alertManager->emplace_alert<alert::InvalidPieceArrived>(_torrent, endPoint, peerId, index);
-
-            // tell session about endpoint and piece
-            _session.invalidPieceReceivedOnConnection(peerId, index);
-
-            // and remove call
-            _outstandingFullPieceArrivedCalls.erase(it);
-        }
-    }
 }
 
 void TorrentPlugin::tick() {
 
     // Asynch processing in session if its setup
-    if(_session.mode() != protocol_session::SessionMode::not_set)
+    if(_session.mode() != protocol_session::SessionMode::not_set) {
         _session.tick();
+    }
+
+    if(_session.mode() == protocol_session::SessionMode::buying &&
```

Comment:
Why is this policing not happening inside `protocol_session` ? Actually, even the `mode` guard could be inside of that same routine?